File: src/main/java/nl/jhvh/java/geometry/model/GeometryType.java
package nl.jhvh.java.geometry.model;

import javax.annotation.Nullable;
import java.util.Arrays;
import java.util.Map;

import static java.util.stream.Collectors.toMap;

public enum GeometryType {

    PARALLELOGRAM("Par"),
    RECTANGLE("Rct"),
    SQUARE("Sq");

    private final String abbreviation;

    GeometryType(String abbreviation) {
        if (abbreviation == null || abbreviation.isEmpty()) {
            throw new IllegalArgumentException("abbreviation of a " + GeometryType.class.getSimpleName() + " must not be null or empty!");
        }
        this.abbreviation = abbreviation;
    }

    // abbreviation must be unique
    // If not unique, mapping would fail with ExceptionInInitializerError - and that's what we want anyway !
    // Caused by: java.lang.IllegalStateException: Duplicate key .. (attempted merging values ... and ...)
    private static final Map<String, GeometryType> typeMapper = Arrays.stream(values())
            .collect(toMap(g -> g.abbreviation, g -> g));

    @Nullable
    public static GeometryType byAbbreviation(String abbreviation) {
        return typeMapper.get(abbreviation);
    }

}

File: src/main/kotlin/nl/jhvh/kotlin/geometry/model/GeometryType.kt
package nl.jhvh.kotlin.geometry.model

enum class GeometryType(val abbreviation: String) {
    PARALLELOGRAM("Par"),
    RECTANGLE("Rct"),
    SQUARE("Sq");

    companion object {
        // abbreviation must be unique, but this mapping does NOT throw an exception when not unique!
        // If we really want so, we have to check uniqueness by ourselves
        private val typeMapper: Map<String, GeometryType> = values().map { it.abbreviation to it }.toMap()

        init {
            // Throws IllegalStateException when not meeting condition
            check(typeMapper.size == values().size) {"Duplicate key in ${
                values().map { "${it.abbreviation} -> $it" } }"}
        }

        fun byAbbreviation(abbreviation: String): GeometryType? = typeMapper[abbreviation]
    }
}